backend/server.js

```diff
@@ -274,11 +359,70 @@ app.post("/users/:id/change-password", (req, res) => {
     .then(() => {
       res.status(201).json({ message: "Käyttäjän salasana tallennettu" });
     })
-    .catch((error) => {
+    .catch(() => {
       res.status(500).json({ message: "Virhe!" });
     });
 });
 
+async function validateToken(req, res, next) {
+  const idToken = req.headers.authorization?.split("Bearer ")[1];
+  if (!idToken) {
+    return res.status(401).json({ message: "Unauthorized" });
+  }
+
+  try {
+    const decodedToken = await getAuth().verifyIdToken(idToken);
+    req.user = decodedToken;
+    next();
+  } catch (error) {
+    return res.status(401).json({ message: "Invalid token" });
+  }
+}
+app.post("/users/:id/permissions", validateToken, async (req, res) => {
+  if (req.user.admin !== true) {
+    return res.status(401).json({ message: "Unauthorized" });
+  }
+
+  const uid = req.params.id;
+  const userRecord = await getAuth().getUser(req.user.uid);
+  const checkboxes = req.body.checkboxes;
+  const reports = [
+    "Kaikki tunnit kumulatiivinen",
+    "Kaikki tunnit",
+    "Laskutettavat tunnit",
+    "Palkka",
+  ];
+  const customClaimsReports = {};
+  let customClaims = {};
+  reports.forEach((item) => {
+    if (checkboxes.includes(item)) {
+      customClaimsReports[item] = true;
+    } else {
+      customClaimsReports[item] = false;
+    }
+  });
+
+  if (userRecord.customClaims.admin) {
+    const adminClaims = userRecord.customClaims;
+    customClaims = { ...adminClaims, ...customClaimsReports };
+  } else {
+    customClaims = { ...customClaimsReports };
+  }
```

Comment:
Instead of this we could just get users current custom claims and add the report claims there, so:
`customClaims = { ...userRecord.customClaims, ...customClaimsReports };`
That way it's not only keeping previous claims when the user is admin. For example if a new claim, "manager" is added, it would be automatically kept in claims even though report permissions are modified.